### ISSUE TITLE
Problem when putting another tag before a param tag in a group

### ISSUE THREAD

**Issue by phil-davis** [squizlabs/PHP_CodeSniffer#1905]
Make a doc block like this:
```
	/**
	 * Some text about this step/method
	 *
	 * @param string $name
	 * @return void
	 * @When I create a file called :name
	 */
```

``phpcs``, for the ``When`` line, spits out the error ``Tag cannot be grouped with parameter tags in a doc comment``
which is correct - if I like, I can suppress this in the ``standards`` settings, and put all the tags without blank lines between each tag-type.

Move the ``When`` tag above the ``param`` tag:
```
	/**
	 * Some text about this step/method
	 *
	 * @When I create a file called :name
	 * @param string $name
	 * @return void
	 */
```

``phpcs``, for the ``param`` line, spits out the error ``Parameter tags must be grouped together in a doc comment``

That is incorrect - the ``param`` lines (in this case only 1 ``param`` line)  are "grouped together". The problem is that there is a non-param tag in the group.

It should spit out, for the ``When`` line, the error ``Tag cannot be grouped with parameter tags in a doc comment``

The code processing this case is not understanding that the non-param tag could appear in a``param`` group before the first ``param`` tag line is found in the group.

**Comment by phil-davis** [comment]
I noticed this because I want to suppress ``Tag cannot be grouped with parameter tags in a doc comment`` and keep reporting ``Parameter tags must be grouped together in a doc comment``

But if I have that combination and write a doc block like the 2nd example above, then I get a fail when it should pass.

**Comment by gsherwood** [comment]
> That is incorrect - the param lines (in this case only 1 param line) are "grouped together". The problem is that there is a non-param tag in the group.

When the sniff says "grouped together" it means there should be a blank line before and after the grouping of param tags. Like this:
```php
/**
 * Some text about this step/method
 *
 * @When I create a file called :name
 *
 * @param string $name
 *
 * @return void
 */
```
That's what the sniff is trying to enforce, so the error is correct.

**Comment by phil-davis** [comment]
I want to suppress ``Tag cannot be grouped with parameter tags in a doc`` so in my ``phpcs.xml`` is:
```
	<exclude name="Generic.Commenting.DocComment.NonParamGroup" />
```
and then I [want|expect] any of the following doc blocks to pass:

Example 1:
```
	/**
	 * Some text about this step/method
	 *
	 * @param string $name
	 * @param string $folder
	 * @return void
	 * @When I create a file called :name in :folder
	 */
```

Example 2:
```
	/**
	 * Some text about this step/method
	 *
	 * @param string $name
	 * @param string $folder
	 * @When I create a file called :name
	 * @return void
	 */
```

Example 3:
```
	/**
	 * Some text about this step/method
	 *
	 * @When I create a file called :name
	 * @param string $name
	 * @param string $folder
	 * @return void
	 */
```

Examples 1 and 2 do pass. But example 3 gives the error ``Parameter tags must be grouped together in a doc comment``

All 3 examples are the same "form" - allowing all the tags to be in a block together. Currently for this case the sniff [expects|enforces] that the ``param`` tag(s) [is|are] first in the block of tags (somewhat accidentally I believe) and if not, emits ``Parameter tags must be grouped together in a doc comment``

**Comment by gsherwood** [comment]
> I want to suppress Tag cannot be grouped with parameter tags in a doc so in my phpcs.xml is:

That's not how this sniff works. The sniff generates multiple errors, and you can exclude the ones you want, but it doesn't change the behaviour of the sniff. It's still going to report those other grouping errors that you don't want.

What you probably need is the sniff to get a property setting, which allows you to change its behaviour. But this sniff doesn't have any, so I don't think it's going to work the way you want it to.

**Comment by phil-davis** [comment]
Oh well - in my opinion the behaviour of ``Parameter tags must be grouped together in a doc comment`` is inconsistent, and I am just fixing that.

But if you do not agree, then I guess that is that.

**Comment by gsherwood** [comment]
> Oh well - in my opinion the behaviour of Parameter tags must be grouped together in a doc comment is inconsistent,

I don't understand why you think it is inconsistent. The sniff is trying to ensure you have a blank line before and after the `@param` tag block, so it is throwing errors if that condition not met.

**Comment by phil-davis** [comment]
The 3 examples all do have a blank line before and after the whole block that contains a group of ``param`` tags. The actual ``param`` tags can have other tags before/after them and the block as a whole has a blank line before/after.

**Comment by gsherwood** [comment]
> The 3 examples all do have a blank line before and after the whole block that contains a group of param tags.

That's not what I'm saying. The sniff wants the @param group to only contain @param tags, and to have a blank line before and after. Like this:
```php
    /**
     * Returns the content of the tokens from the specified start position in
     * the token stack for the specified length.
     *
     * @param int $start  The position to start from in the token stack.
     * @param int $length The length of tokens to traverse from the start pos.
     *
     * @return string The token contents.
     */
```
That specific format is what the included PEAR and Squiz coding standards want, and so that's how the sniff has been coded. It does not have any settings to change this behaviour.

**Comment by phil-davis** [comment]
Well, respectfully, the current code does allow for variations, examples 1 and 2 do not emit ``Parameter tags must be grouped together in a doc comment``

They correctly, IMO, emit ``Tag cannot be grouped with parameter tags in a doc comment`` - which the actual reason for them not meeting the standard.

If example 3 would also emit ``Tag cannot be grouped with parameter tags in a doc comment`` as the error, (rather than ``Parameter tags must be grouped together in a doc comment``) then I am happy.

That is what I see as the inconsistent behaviour.

All 3 examples are "wrong" (they do not meet the standard), and all do not meet it for the same reason.

**Comment by gsherwood** [comment]
It took me a while, but I think I see what you're asking for now. Thanks for continuing to post until it clicked. I'll take a look into this a bit more, and that PR you sent.

**Comment by gsherwood** [comment]
Closed by PR #1906
